File: BOJ/17128.cpp
#include<bits/stdc++.h>
using namespace std;

int quality[200001];
int cow[200001];

int main() {
	ios_base::sync_with_stdio(false); // 두 줄을 쓰지 않으면 시간 초과가 발생한다.
	cin.tie(nullptr);

	int N, Q;
	cin >> N >> Q;
	for (int i = 1; i <= N; i++) {
		cin >> cow[i];
	}

	for (int i = 0; i < 200001; i++) quality[i] = 1; // 원소들을 곱해줘야 하기 때문에 1로 초기화

	int sum = 0;
	for (int i = 1; i <= N; i++) {
		for (int j = i; j <= i + 3; j++) { // quality[i]는 문제에서 주어진 S를 구하기 위한 A_i * A_i+1 * A_i+2 * A_i+3을 구현.
			if (j%N == 0) quality[i] *= cow[j]; // 0번째 index = N번째 index
			else quality[i] *= cow[j%N];
		}
		sum += quality[i];
	}

	for (int i = 0; i < Q; i++) {
		int idx;
		cin >> idx;
		for (int j = idx - 3; j <= idx; j++) {
			if (j%N == 0) { // 0번째 index = N번째 index
				quality[N] *= -1;
				sum += 2 * quality[N];
			}
			else if (j%N < 0) { // 음수이면 N+(j%N)번째 인덱스가 된다. ex) -1 -> N, -2 -> N-1
				quality[N + (j%N)] *= -1;
				sum += 2 * quality[N + (j%N)];
			}
			else {
				quality[j%N] *= -1;
				sum += 2 * quality[j%N];
			}
		}
		cout << sum << '\n';
	}

	return 0;
}

File: BOJ/python/17128.py
import sys

input = sys.stdin.readline

N, Q = map(int, input().rstrip('\n').split(' '))
quality = list(map(int, input().rstrip('\n').split(' ')))
joke = list(map(int, input().rstrip('\n').split(' ')))
calc = [1 for _ in range(len(quality))]

# i번째 S = A_i * A_i+1 * A_i+2 * A_i+3
for i in range(N):
    calc[i] = quality[i % N] * quality[(i + 1) % N] * quality[(i + 2) % N] * quality[(i + 3) % N]

total = sum(calc)

for idx in joke:
    for i in range(idx - 3, idx + 1):
        calc[(i - 1) % N] *= -1 # python은 음수 index가 가능하기 때문에 나머지를 그냥 넣어줘도 된다.
        total += 2 * calc[(i - 1) % N]
    print(total)
